### PR TITLE
feat: pass retry from Job.result() to Job.done()

### PR DESCRIPTION
**Pull request by IlyaFaer**
Closes https://github.com/googleapis/python-bigquery/issues/24

### CODE REVIEW

**Comment by IlyaFaer** [comment]
This PR is strictly connected to https://github.com/googleapis/python-api-core/pull/9, and should not be merged before №9

**Comment by IlyaFaer** [comment]
Here is how it's gonna be passed into `done()`

image

**Comment by tswast** [comment]
This will need to be updated to set the minimum api-core version in setup.py to 1.23.0 after https://github.com/googleapis/python-api-core/pull/96 is released.

**Comment by tswast** [comment]
1.23.0 has been released. Work can continue on this PR

**Comment by IlyaFaer** [comment]
Only one error is failing:
```
self = <google.cloud.bigquery.job._AsyncJob object at 0x7f1fc2f0eb70>

    def to_api_repr(self):
        """Generate a resource for the job."""
>       raise NotImplementedError("Abstract")
E       NotImplementedError: Abstract
```
Doesn't seem to be related.

File: google/cloud/bigquery/job.py [hunk 1]
```diff
@@ -817,8 +817,9 @@ def result(self, retry=DEFAULT_RETRY, timeout=None):
         """
         if self.state is None:
             self._begin(retry=retry, timeout=timeout)
-        # TODO: modify PollingFuture so it can pass a retry argument to done().
-        return super(_AsyncJob, self).result(timeout=timeout)
+
+        kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
```
Comment (tswast) [inline google/cloud/bigquery/job.py:821]: We don't have to follow this pattern here. We just need to bump the minimum `google-api-core` dependency version in `setup.py` and our `constraints-X.txt` file(s).

**Review by tswast — CHANGES_REQUESTED** [review]
Missing update to minimum versions.

**Comment by tswast** [comment]
Filed https://github.com/googleapis/python-bigquery/issues/339 to look into the test failure. I've assigned to myself, as it might be related to some refactoring I recently completed.

**Comment by IlyaFaer** [comment]
I think, a lot of tests will fail. And in many tests we'll have to add one more arg. That's why I've added the `kwargs` pattern.

**Comment by tswast** [comment]
> I think, a lot of tests will fail. And in many tests we'll have to add one more arg. That's why I've added the kwargs pattern.

Okay, that makes sense. So long as we have some tests that have a custom retry, I'm okay with only populating it when it's something other than the default.

**Comment by HemangChothani** [comment]
@googlebot I consent

**Comment by IlyaFaer** [comment]
@tswast, sorry, I'm little bit besieged by tasks, not able to continue with this right now. @HemangChothani kindly agreed to take a look at this PR.

File: tests/unit/test_job.py [hunk 2]
```diff
@@ -4812,6 +4811,58 @@ def test_result_with_max_results(self):
...
+        job_resource_done["configuration"]["query"]["destinationTable"] = {
+            "projectId": "dest-project",
+            "datasetId": "dest_dataset",
+            "tableId": "dest_table",
+        }
+
+        connection = _make_connection(
+            exceptions.NotFound("not normally retriable"),
+            query_resource,
+            query_resource_done,
```
Comment (tswast) [inline tests/unit/test_job.py:4838]: We had a few places where custom retry was already supported. I'd like to see a few more exceptions to be totally sure that all API calls are getting the right retry object.

```suggestion
            exceptions.NotFound("not normally retriable"),
            query_resource_done,
            exceptions.NotFound("not normally retriable"),
```

**Review by tswast — CHANGES_REQUESTED** [review]

**Review by tswast — APPROVED** [review]

**Pull request merged**